src/halmos/utils.py

```diff
@@ -297,9 +297,6 @@ def unbox_int(x: Any) -> Any:
     if is_bv_value(x):
         return x.as_long()
 
-    if is_bv(x):
-        x = simplify(x)
-
```

Comment:
I think it's weird for `unbox_int` to try to convert to an int, but _also_ simplify if it's not.

I think simplify should happen at well defined boundaries, like:
- in and out of the stack
- in and out of memory
- *not* in an intermediate place like unbox_int

> unbox_int(ex.context.output.data) == ASSERT_FAIL

☝️ this is not a great use of unbox_int, if the result is a bv we produce a new cond rather than evaluating a bool. In this case we probably want to call simplify first, then try to unbox, and skip if the unboxed output is still a bv

Reply:
> the result of get_word() is better to be simplified, as it is used in many cheatcodes, and passed through further logic

I agree, that's why the output of `unwrap` is simplified

Reply:
> In this case we probably want to call simplify first, then try to unbox, and skip if the unboxed output is still a bv

i agree

> I agree, that's why the output of unwrap is simplified

thanks for clarification, now i see that it's already simplified before being passed into unbox_int

--

i agree that it would make more sense to not simplify inside unbox_int. then, let me double-check other places to make sure:

- https://github.com/a16z/halmos/blob/c3f45dd3ff65635840847713ece60b779f8600d5/src/halmos/sevm.py#L727-L728 this is fine because `self.pgm[self.pc]` is already simplified, right?

- https://github.com/a16z/halmos/blob/c3f45dd3ff65635840847713ece60b779f8600d5/src/halmos/utils.py#L265-L269 if i understand correctly, `calldata[:4]` may not be simplified, right? for now, it's fine because extract_funsig is immediately passed to `int_of`, which would revert anyway if it's a bv. (there is another use of extract_funsig, but it will be removed anyway in #349, so it doesn't matter.) however, for the longer term, should we simplify `calldata[:4]` here? wdty?

Reply:
> could you please fix that as you suggested? 

yea 👍

> self.pgm[self.pc] is already simplified, right?

it defers to `ByteVec.get_byte()`, which calls `chunk.get_byte()`
- for a concrete chunk, we get an int
- for a symbolic chunk, it relies on `extract_bytes`, which does return simplified bitvecs

so rn yes, but it's not strongly part of the contract (it just so happens that extract_bytes does simplify)

> should we simplify calldata[:4] here? 

yes, instead of "trying" to unbox it, returning `calldata[:4].unwrap()` would do it (return either a concrete value or a simplified bv)

Reply:
> could you please fix that as you suggested?

more robust version in af3df48 and test in afd5c11q. I confirmed that `unbox_int(...) == ASSERT_FAIL` does produce a BoolRef of the form `Concat(1313373041, p_x_uint256) == 152078208365357342262005707660225848957176981554335715805457651098985835139029979365377`, which for some reason evaluates to False in the condition (instead of generating an exception as I was expecting).

Reply:
> instead of "trying" to unbox it, returning calldata[:4].unwrap() would do it (return either a concrete value or a simplified bv)

great, let's do it!